Add a trigger field to the GCM push notification payload

kunall17 opened the pull request:
Used some constant strings to denote the event instead of using the whole trigger object to save some space for the payload
```
message.triggers = {
            'private_message': False,
            'mentioned': False,
            'stream_push_notify': True,
}
```

And haven't tested the newly added condition in a server but should work!

timabbott commented on zerver/lib/push_notifications.py:
```diff
@@ -322,14 +322,34 @@ def remove_push_device_token(user_profile, token_str, kind):
+            (message.triggers['stream_push_notify'] and message.stream_name)):
+        return "stream"
+    else:
+        return "multiple"
```
I actually had been thinking about changing the structure of the original `triggers` data set to just be the single highest priority trigger in the backend, i.e. just passing private, mentioned, or stream_push_notify as the trigger.  What do you think about that? 

I guess my thinking is that it's not actually interesting whether or not a user was mentioned in a group PM, and that would make the protocol a lot simpler.

kunall17 replied:
This is the original data you're referring to right https://github.com/zulip/zulip/blob/master/zerver/lib/push_notifications.py#L406 ?

So for this data, yeah I also think this should be simplified to string constants and sent directly.

> I guess my thinking is that it's not actually interesting whether or not a user was mentioned in a group PM, and that would make the protocol a lot simpler.

I think, if we are giving the user information about mentioning on a stream, then we should give the information about a mention on the group, and it's really just adding a trigger constant so it's the work required is also very less!

kunall17 replied:
@timabbott What about this one?

timabbott commented:
What I'd like to do is change this block:
```
    if (idle or always_push_notify) and (private_message or mentioned or stream_push_notify):              
        notice = build_offline_notification(user_profile_id, message_id)                                   
        notice['triggers'] = {                                                                             
            'private_message': private_message,                                                            
            'mentioned': mentioned,                                                                        
            'stream_push_notify': stream_push_notify,                                                      
        }                                                                                                  
```

to set `trigger` to be one of `private_message`, `mentioned`, or `stream_push_notify` (based on which is the first of those settings that's set) and then have the logic based on those.  You could take a crack at doing that.

kunall17 commented:
And two more strings `group_message` & `group_mentioned` ?

kunall17 commented on zerver/tornado/event_queue.py:
```diff
@@ -657,20 +658,29 @@ def receiver_is_off_zulip(user_profile_id):
-            'stream_push_notify': stream_push_notify,
-        }
+        
+        if (group_message and mentioned):
```
These strings to cover all the possible cases?

timabbott replied:
Is there a reason to distinct `group_message` from `private_message` here?

kunall17 replied:
Basically I wanted to create strings for every possible case, 
If we want to reduce strings we can either go for `notice['mentioned'] = false` and check Recipient.[CASE] and then we don't need triggers.

kunall17 replied:
But in the client side, there is no distinction b/w PM or a group message, as the recipient_type has `stream | private` so it could be better to include everything in the trigger field (if no other solution found)

kunall17 commented:
@timabbott If I change recipient_type to include `stream|private|group` rather than `stream|private` here, the whole code can be simplified a lot, as the main problem is the client doesn't have distinction b/w group & private!
Will it be fine to include it? 

Currently I'm wondering how to determine whether it's a group message in `maybe_enqueue_notifications_for_message_update` as it uses `maybe_enqueue_notifications` and I'm passing group_message to `maybe_enqueue_notifications` like the `private_message`

timabbott commented:
@kunall17 it looks like this has a lot of test failures; can you try cleaning it up?

I think it'd be easier if you split this into 2 commits.  The first would change the format of `triggers` -> `trigger`, without changing any of the existing cases; that should be easy to test/review/merge.  And the second can add the new cases that you're interested in, which I'm more uncertain about the details of.

It looks like the old Android app's push notifications for group PMs will break if we change `recipient_type`.  But the webapp doesn't use it, so it might be possible to change it if we were to do a release of the old Android app to support it.  Not sure if that's worth it.

kunall17 commented:
Yeah adding a group string to the recipient_type will have problems in the sub text for the old android app.

I'll do the splitting and changes after Tuesday (exams going on)

timabbott commented:
I just merged e98ca0714b36e92011488a9879909e9ccbbeeeba, which is basically the split-out first commit we'd been discussing.

kunall17 commented:
@timabbott Updated

timabbott commented:
@kunall17 I don't really understand what the goal of this set of changes is; it seems like the messages are going to be the same as before?

kunall17 commented:
The trigger is being sent which will be used for correct grouping in the notifications (once they start stacking up) 
Currently there was no way to determine whether the notification is for the group or group mention or private message.

kunall17 commented:
@timabbott Updated PR, it uses mentioned instead of trigger field in the missed_message to compute the new trigger supporting the new trigger types which are required in the client side to properly handle the count of notifications

gnprice commented:
@kunall17 I don't understand what the purpose of this PR is; can you explain?

This is doing at least two different things which should definitely be separate commits:
* In the internal event, it changes from sending `trigger` to sending just a boolean `mentioned`, and moves the logic to identify the trigger from Tornado over to the queue processor.
* It changes what that logic does, adding group mentions.
* As it changes the logic that identifies the trigger for the notification, it might be changing what happens in some cases other than group mentions; I'm not sure.

The first change is basically the reverse of what Tim did in commit e98ca0714 (mentioned above), so it seems like it might be the wrong direction architecturally.

kunall17 commented:
The main objective of this PR is to include a trigger field in the GCM payload which would help in detection of narrow for this notification of message, currently it is not possible to distinguish b/w private/group/group mention & b/w stream/stream mention.

kunall17 commented:
This as an addition to some cases for Tim's commit, as his commit converted the previous dict of the trigger into a string.

Previously the trigger was being passed through the tornado event_queue.py now I'm sending a mentioned boolean which was hard to determine in the push_notifications, And using the mentioned and seeing the narrow of the message a trigger string is being constructed. 

Should I move the trigger construction in the tornado event queue it?

gnprice commented:
Let's start by adding the information you want from the Tornado side, where we compute `trigger` now. If that gets complicated, or there's some expensive computation you want to do that shouldn't happen in Tornado, then we can move it over to the queue-worker side. Either way, let's definitely keep to separate commits for any moving of where the logic is and any changes to what it does.

kunall17 commented:
In (maybe_enqueue_notifications)  https://github.com/zulip/zulip/blob/master/zerver/tornado/event_queue.py#L667 basically, we only need a boolean whether it's a group or a private message rest all the information we have there
To determine this in the queue-worker it uses 

`(message, user_message) = access_message(user_profile, missed_message['message_id'])`
 
From the `message` we determine it's narrow. 

I'll split out the current commits as well!

gnprice commented:
This splitting doesn't work -- after each intermediate commit, the code is broken. We always want each commit to leave the code in a reasonable and fully-working state.

How about for a first commit, just change `maybe_enqueue_notifications` to compute the same set of `trigger` values that you compute in your `create_trigger_from_message`? Then we'll be able to look closely at what cases that causes things to change in.

kunall17 commented:
@gnprice Made changes, squashed the commits, the code is not broken now & each commit is in fully working state, just the test cases are added in the last commit, will it be fine or should I split them as well?

gnprice commented:
Take a look at your first commit, 63407e2d7: it changes the values that `message.trigger` can have (e.g. to say `stream_message` instead of `stream_push_notify`), but it doesn't update the logic in `get_alert_from_message` that uses those values. So `get_alert_from_message` can't do its job properly -- the code is broken.

And the tests do fail -- check out that commit and run them. That's a good thing, because if no tests failed in that case then that would mean we didn't have adequate tests on this code. :wink: 

We always want each commit to leave the code in a state where both (a) the code actually works, and (b) the tests all pass. If we don't have (b), it becomes very hard to have confidence in (a). And if we don't have (a), it becomes very hard to understand the changes, because the individual commits don't make sense on their own, and because the reader can't reason based on what would need to happen in order for things to work -- because they don't work.

At the same time, the commit does two different things that shouldn't be coupled together in the same commit: it moves where the logic for computing `trigger` happens, from the Tornado process to the queue worker, at the same time as it changes what that logic does. That makes the commit harder to understand, both in review and for someone studying it later.

Take another read through https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-discipline . It's important.

kunall17 commented:
Ok, now I understood the requirements, thanks for the explanation.

So how about these commits? 

1. Computes the trigger in the tornado, having triggers as ( private_message, stream_message, stream_mention ) as the group/private is not distinguishable on the tornado side | fix the test's 

1. Pass this computation of trigger to the queue worker side & send the trigger with GCM & APNS

1. Add the group trigger's detection in the queue worker side

If we agree I'll make the code changes?

gnprice commented:
> as the group/private is not distinguishable on the tornado side

OK, great, this gets to the core of it. Why is this not distinguishable on the tornado side?

kunall17 commented:
Currently, this method has these params

 `maybe_enqueue_notifications(user_profile_id, message_id, private_message,
                                mentioned, stream_push_notify, stream_name,
                                always_push_notify, idle, already_notified):`

and only `message_id` would tell me after searching in the db and I don't want to change the params of this method which I believe will be a performance overhead as the queue worker has all the info! 
 
If there is any other method to determine, please let me know!

gnprice commented:
This is a tricky part of the system, so this is a good discussion to have. I'm not an expert in this area myself, so I'm reading up on it just like you are.

You're quite right that we definitely don't want to try to fetch from the DB in Tornado -- that code has to rely on the information it's given.

At the call sites of `maybe_enqueue_notifications`, the information we have is what's in the event sent to Tornado, and it's an event of type `message`. Searching the codebase, that comes from this statement in `zerver/lib/actions.py` (with a few more fields added later):
```
        event = dict(
            type='message',
            message=message['message'].id,
            message_dict=wide_message_dict,
            presence_idle_user_ids=presence_idle_user_ids,
        )
```
The most interesting of these is `message_dict`. Tracing through the code a bit, that ultimately comes from `to_dict_uncached_helper` in `zerver/lib/message.py`, and its helper `build_message_dict`. There's actually a lot of detail in there! Take a look. This dict isn't called "wide" for nothing. :grin: 

Within that `message_dict`, I think the key item is `recipient_type`. This is going to be `Recipient.PERSONAL` for a one-to-one PM, and `Recipient.HUDDLE` for a group PM. So that should give you what you need.

... Hm. There actually is one case (out of the 3 call sites for `maybe_enqueue_notifications`) where we don't have a `message` event -- we have an `update_message` event. And those don't contain a message dict like this. ... OK, but that case (the one in `maybe_enqueue_notifications_for_message_update`) only happens for stream messages, so we're OK because we can just ignore this distinction there.

So I think by using `event['message_dict']['recipient_type']`, you can get the information you need. Want to try that?

Do ask questions about any of this that doesn't make sense after a little poking at it.

kunall17 commented:
Ok, last time I tried to edit the `maybe_enqueue_notifications` crashed a lot of test cases, so I was a little bit reluctant to mess with the params, but I'll try that again. Thanks for the info! :)

kunall17 commented:
So the function `def missedmessage_hook(user_profile_id: int, client: ClientDescriptor, last_for_client: bool)` didn't have the recipient_type field, I tried to print the dict (event) in the 4455e80 and the dict came out to be 
```
{
  'id': 0,
  'stream_push_notify': False,
  'type': 'message',
  'message': {
    'id': 31,
    'type': 'private',
    'client': 'test suite',
    'sender_realm_str': 'zulip',
    'content_type': 'text/html',
    'sender_full_name': 'Iago',
    'subject': '',
    'timestamp': 1513230803,
    'avatar_url': None,
    'display_recipient': [{
      'id': 4,
      'full_name': 'King Hamlet',
      'is_mirror_dummy': False,
      'short_name': 'hamlet',
      'email': 'hamlet@zulip.com'
    }, {
      'id': 5,
      'full_name': 'Iago',
      'is_mirror_dummy': False,
      'short_name': 'iago',
      'email': 'iago@zulip.com'
    }, {
      'id': 23,
      'full_name': 'Zulip Feedback Bot',
      'is_mirror_dummy': False,
      'short_name': 'feedback',
      'email': 'feedback@zulip.com'
    }],
    'reactions': [],
    'sender_id': 5,
    'recipient_id': 29,
    'content': '<p><span class="user-mention" data-user-email="hamlet@zulip.com" data-user-id="4">@King Hamlet</span> what\'s up?</p>',
    'subject_links': [],
    'sender_email': 'iago@zulip.com',
    'is_me_message': False,
    'sender_short_name': 'iago'
  },
  'email_notified': True,
  'push_notified': True,
  'flags': ['mentioned']
}
```

I obtained this by editing the test cases and added a send_huddle_message test case for confirming everything (All these changes are in the commit 4455e80) 

So will it be fine to add this field to the queue creation side? (Not even sure how's that done)

gnprice commented:
This is complex and has been inactive for a long time, and the original motivating issue zulip/zulip-mobile#1194 has been resolved.

There may still be something for us to do in this direction. If so, when we come to that we'll make a fresh implementation.
